### PR TITLE
#82

### PR DESCRIPTION
**Pull request by 0leksandrr**
## Changes Made
- **Filtering by Name**: 
  - Added a filter option that enables users to retrieve startup profiles based on their names. 
  - This functionality enhances user experience by allowing targeted searches for specific startups.

- **Search by Description**: 
  - Implemented a search feature that allows users to find startup profiles by entering keywords related to their descriptions.
  - This helps in discovering relevant startups quickly.

- **Partial Name Search**: 
  - Introduced the capability to search for startups by partial names, providing more flexibility in locating profiles.

- **Tests**: 
  - Created a comprehensive set of tests to validate the new filtering and search functionalities.
  - Ensured that all edge cases are covered, including filtering and searching for both exact and partial matches.

### CODE REVIEW

File: startups/filters.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+class StartUpProfileFilter(django_filters.FilterSet):
+    """
+    name: Filters the startup profiles by name.
+    description: Filters the startup profiles by description.
+    created_at: Filters the startup profiles by creation date range.
+    """
+
+    name = django_filters.CharFilter(field_name='name', lookup_expr='icontains', label='Startup Name')
+    description = django_filters.CharFilter(field_name='description', lookup_expr='icontains', label='Description')
+    created_at = django_filters.DateFromToRangeFilter(field_name='created_at', label='Creation Date Range')
```
Comment (mehalyna) [inline startups/filters.py:14]: If invalid date formats are provided for `created_at` filters, it may raise a `ValidationError` and cause unhandled exceptions.
You may implement **custom validation** to catch such errors gracefully and return user-friendly error messages.
```
def filter_created_at(self, queryset, name, value):
    try:
        return queryset.filter(created_at__range=value)
    except (ValueError, TypeError):
        raise django_filters.exceptions.ValidationError("Invalid date range format.")
```

File: startups/tests.py [hunk 2]
```diff
@@ -1,19 +1,65 @@
...
+            password="123456pok",
+            first_name="John",
+            last_name="Doe",
+            user_phone="+1234567890"
+        )
+        self.user.add_role("Investor")
+
+        StartUpProfile.objects.create(user_id=self.user, name="Tech Solutions", description="Innovative tech startup")
+        StartUpProfile.objects.create(user_id=self.user, name="Eco Start", description="Sustainable energy startup")
+        StartUpProfile.objects.create(user_id=self.user, name="Foodie", description="Food delivery service startup")
```
Comment (mehalyna) [inline startups/tests.py:24]: You are assigning the user using the field `user_id`. It's recommended to use the related field name directly (e.g., `user=self.user`) for better clarity and readability.

File: startups/tests.py
Comment (mehalyna) [inline startups/tests.py:1]: It’s not clear whether your filtering/searching mechanism is case-insensitive. You should ensure that a search for "tech solutions" (lowercase) returns the correct result even if the name is "Tech Solutions".
Add a test case for case-insensitive searches

File: startups/tests.py [hunk 3]
```diff
@@ -1,19 +1,65 @@
...
+
+    def authenticate_user(self):
+        url = reverse('jwt-create')
+        response = self.client.post(url, {
+            'email': 'john@gmail.com',
+            'password': '123456pok',
+        })
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        token = response.data['access']
+        self.client.credentials(HTTP_AUTHORIZATION=f'Bearer {token}')
```
Comment (mehalyna) [inline startups/tests.py:38]: The `authenticate_user` method is separated from `setUp`. For clarity and simplicity, you can move the authentication logic directly into `setUp` since it’s always needed for the tests.

File: startups/tests.py
Comment (mehalyna) [inline startups/tests.py:1]: There are no test cases to ensure the behavior of the filter when no results match the query.
Add a test case for non-matching filters

Comment (mehalyna) [inline startups/tests.py:1]: There are no tests for empty search queries or invalid inputs.
Add tests to ensure the API handles such cases gracefully

File: startups/views.py [hunk 4]
```diff
@@ -14,25 +16,42 @@
...
     permission_classes = [IsAuthenticated]
 
     def create(self, request, *args, **kwargs):
         response = super().create(request, *args, **kwargs)
-        return get_success_response('Startup profile created successfully', response.data, status.HTTP_201_CREATED)
+        return get_success_response(
+            message='Startup profile created successfully',
+            data=response.data,
+            status=status.HTTP_201_CREATED
+        )
```
Comment (mehalyna) [inline startups/views.py:48]: Wrap the call to `super().create()` in a try-except block to handle potential errors gracefully.

File: startups/tests.py
Comment (mehalyna) [inline startups/tests.py:1]: Your API uses pagination. Your test cases might fail because only a subset of results will be returned.
You can add assertions for pagination metadata.

**Review by mehalyna — CHANGES_REQUESTED** [review]

File: startups/filters.py (hunk 1 above)
Comment (0leksandrr) [reply startups/filters.py:14]: done

File: startups/tests.py (hunk 3 above)
Comment (0leksandrr) [reply startups/tests.py]: done

File: startups/tests.py
Comment (0leksandrr) [reply startups/tests.py:1]: done

File: startups/views.py (hunk 4 above)
Comment (0leksandrr) [reply startups/views.py]: done

File: startups/tests.py
Comment (0leksandrr) [reply startups/tests.py:1]: done

File: startups/tests.py (hunk 2 above)
Comment (0leksandrr) [reply startups/tests.py:24]: I have to rename the user_id field to just user?

**Review by mehalyna — APPROVED** [review]

**Review by PolinaCh14 — APPROVED** [review]

**Review by VIad0s — APPROVED** [review]
